### PR TITLE
Issue 8550: Check for a good table_definition_cache value

### PR DESCRIPTION
**Pull request by annando**
This helps with issue https://github.com/friendica/friendica/issues/8550

We now check for a good value for the `table_definition_cache` since a too low value can cause database errors when working with views.

### CODE REVIEW

File: src/Database/Database.php [hunk 1]
```diff
@@ -1645,6 +1645,21 @@ public function processlist()
...
+	 * @param string $name
+	 * @return string content
+	 */
+	public function getVariable(string $name)
+	{
+		$result = $this->fetchFirst("SHOW GLOBAL VARIABLES WHERE `Variable_name` = ?", $name);
+		if (!isset($result['Value'])) {
+			return null;
+		}
+		return $result['Value'];
```
Comment (MrPetovan) [inline src/Database/Database.php:1660]: Equivalent and shorter:
```suggestion
		return $result['Value'] ?? null;
```

File: src/Module/Admin/Summary.php [hunk 2]
```diff
@@ -55,6 +55,16 @@ public static function content(array $parameters = [])
 			$warningtext[] = DI::l10n()->t('Your DB still runs with InnoDB tables in the Antelope file format. You should change the file format to Barracuda. Friendica is using features that are not provided by the Antelope format. See <a href="%s">here</a> for a guide that may be helpful converting the table engines. You may also use the command <tt>php bin/console.php dbstructure toinnodb</tt> of your Friendica installation for an automatic conversion.<br />', 'https://dev.mysql.com/doc/refman/5.7/en/innodb-file-format.html');
 		}
 
+		// Avoid the database error 1615 "Prepared statement needs to be re-prepared", see https://github.com/friendica/friendica/issues/8550
+		$table_definition_cache = DBA::getVariable('table_definition_cache');
+		$table_open_cache = DBA::getVariable('table_open_cache');
+		if (!empty($table_definition_cache) && !empty($table_open_cache)) {
+			$suggested_definition_cache = max(400, round($table_open_cache / 2, 1));
```
Comment (MrPetovan) [inline src/Module/Admin/Summary.php:62]: How did you decide on this value?

Comment (annando) [reply src/Module/Admin/Summary.php:62]: No, they are described in the linked documentation.

Comment (MrPetovan) [reply src/Module/Admin/Summary.php:62]: In the linked documentation, it only says that the default value is `400 + table_open_cache /2` which isn't the calculation you made here. It doesn't say it's the suggested value either.

Comment (annando) [reply src/Module/Admin/Summary.php:62]: The value shouldn't be lower than that.

Comment (MrPetovan) [reply src/Module/Admin/Summary.php:62]: The minimum value for `table_definition_cache` is `400`, it has nothing to do with the default value which takes `table_open_cache` into account. If you make the following change you don't even have to use `max()`.
```suggestion
			$suggested_definition_cache = 400 + round($table_open_cache / 2, 1);
```

Comment (annando) [reply src/Module/Admin/Summary.php:62]: No, according to the documentation they are using the `max` functionality as well. So it doesn't add, 400 is just the minimum value.

**Comment by AlfredSK** [comment]
When I see that correctly I would get the warning despite I don't have any issue with my setting.
table_open_cache = 2250
table_definition_cache = 860

I would have to raise the table_definition_cache to 1125 to get rid of the warning. Right?

File: src/Module/Admin/Summary.php (hunk 2 above)
Comment (MrPetovan) [reply src/Module/Admin/Summary.php:62]: Can you please show me where? At this point I feel like we're looking at different pages. Here's my reference for what I say above:
Screen Shot 2020-04-29 at 15 35 59
It says `minimum of 400` (included in both our codes), `capped at 2000` (missing in both our codes), and it clearly has a plus sign (`+`) in the default value formula. Please note that it is a default value, not a suggested one either.

**Comment by MrPetovan** [comment]
Yes.

**Comment by annando** [comment]
At our company we have got notification messages that have some "don't show me again" option. I'm unable to code this by myself, but later we could add it.

**Comment by MrPetovan** [comment]
In this case it should set a specific `Config` value (in a specific category for these messages) when you click on "Do not show this to me again" and we need a button to reset all these flags.

**Comment by annando** [comment]
Yes. But this had to be some JS magic, I guess.

**Comment by MrPetovan** [comment]
If we store it in the cookie, yes, otherwise it's a regular Ajax call to a module.

**Comment by annando** [comment]
AJAX is a foreign language I don't speak very well :-)

**Comment by MrPetovan** [comment]
Is it Greek to you? ([Ajax = Greek](https://www.greekmythology.com/Myths/Mortals/Ajax/ajax.html), get it?)

**Comment by MrPetovan** [comment]
Please address https://github.com/friendica/friendica/pull/8563#issuecomment-622435327 so this can be merged.

**Comment by annando** [comment]
I meant it seriously, what I wrote about AJAX. I'm unsure how to do it in a good way.

**Comment by MrPetovan** [comment]
When you click on the "do not show again" button, it triggers a Ajax request to `/admin/dismiss/[notice-identifier]` and sets the Config value `admin-dismiss.[notice-identifier]`. Then in the admin module you check for the value of `admin-dismiss.[notice-identifer]` to run the corresponding test and show the notice.

**Comment by annando** [comment]
Yeah, but how do I do the AJAX call in the best way?

**Comment by MrPetovan** [comment]
```javascript
let notice = $([notice-selector]);
notice.fadeOut(200);
$.post('admin/dismiss/[notice-identifier]')
.fail(function() {
    notice.fadeIn(200);
});
```

**Comment by MrPetovan** [comment]
Please address https://github.com/friendica/friendica/pull/8563#issuecomment-622435327 so this can be merged.

**Comment by annando** [comment]
@MrPetovan this AJAX stuff is more complicated for me than the last pull requests. I can try having a look at it at the weekend.

**Comment by MrPetovan** [comment]
I didn't mean for you to address the AJAX part, but the ideal value calculation part. I'd like to understand how you reached that value given that it's absent from the material you referenced.

**Comment by annando** [comment]
I assumed that the default value is the lowest recommened value. And that formula was described for the default value in the documentation.

**Comment by MrPetovan** [comment]
But you aren't using the exact provided formula, hence my confusion.

**Comment by annando** [comment]
Now the formula is changed to how https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_table_definition_cache says:

`MIN(400 + table_open_cache / 2, 2000)`

Also in the linked description there is the note that the default value is capped at 2000. This formula respects this.

**Comment by MrPetovan** [comment]
Thanks, I'm still not sure if it really can be used as a recommended value but at least it is consistent with documentation now.

**Pull request merged**